### PR TITLE
GLTF: Don't collapse non-joint leaf nodes when importing skeletons

### LINKED ISSUE #75917: Blender asset import creates a new bone for every object parented to same bone
### Godot version

c3ed7af

### System information

Windows 10

### Issue description

When importing a Blender 3D asset the Godot asset importer creates a new bone and a new BoneAttachment3D for each object that was parented to a skeleton bone, e.g. each mesh item. Even if this behavior is wanted there is imo a bug when 100 objects are parented all to the very same bone in Blender the Godot importer adds +100 additional bones instead of just one extra bone.

This can result in a very excessive amount of extra bones on complex skeletons with many attachments for e.g. character customization. As mentioned in related https://github.com/godotengine/godot/issues/75915 having so many extra bones is slow to work with and consumes performance for mostly nothing.

Godot 3 did not add extra bones for each bone attachment but this was apparently changed to solve other bugs with bone index shuffling like https://github.com/godotengine/godot/issues/64512 and https://github.com/godotengine/godot/issues/61641.

### Steps to reproduce

- Open the attached Blender test file (was made with Blender 3.40)
- See both meshes are parented to the same bone.
- Copy blend file to Godot and let importer run.
- See that Godot created two new extra bones that did not exist in the original skeleton.

### Minimal reproduction project

[BlenderTestFile.zip](https://github.com/godotengine/godot/files/11195011/BlenderTestFile.zip)

### LINKED ISSUE #92447: Importing GLTF add fake bones to the skeleton
### Tested versions

- Reproducible in: v4.2.2.stable.official [15073afe3]

### System information

Windows 11 - Godot v4.2.2.stable - Windows 10.0.22631 - GLES3 (Compatibility) - GeForce GTX 1050 Ti () - AMD Ryzen 5 3600 6-Core Processor (12 Threads)

### Issue description

During the import process, when there is an object that is not a joint in the skeleton hierarchy, a fake bone is created in its place, altering the original skeleton hierarchy. Therefore, it is not possible to correctly recover the position of the animated vertex because the bone indexes in the vertex data refer to the original skeleton.

Captura de tela 2024-05-27 163333

Recording 2024-05-27 at 16 38 39

**Temporary solution**

Create an array that maps the indices of the original skeleton to the modified skeleton.

### Steps to reproduce

1. Uncheck "Use Correction" at "debug.gd" attached to "beast/Armature/Skeleton3D/body"
2.  Play the "ArmatureAction" animation in the Animation Player

### Minimal reproduction project (MRP)

[skin_collision.zip](https://github.com/godotengine/godot/files/15461794/skin_collision.zip)

### PR DESCRIPTION
**Pull request by aaronfranke**
Fixes issue #75917, fixes issue #92447, supersedes PR #92452, supersedes PR #98516.

This PR fixes an issue encountered by the Blender Project DogWalk team where non-joint leaf glTF nodes were getting always collapsed into the skeleton instead of being imported as nodes. The file they exported from Blender was correctly able to be imported back into Blender, but did not work correctly in Godot. This PR fixes that.

When I initially opened this PR I tried to keep the changes minimal, but I kept running into edge case after edge case, and I finally decided I needed to completely refactor this code to make it work.

In `SkinTool`, the existing code essentially dug down into the node hierarchy of a skeleton root and assigned everything inside as a joint, but this isn't what we want. The correct behavior is that only joint nodes and their ancestors up to the skeleton root should be generated as bones, along with the first level of leaf bones. If a leaf glTF node is not a joint, we should allow it to stay as a node, following the skeleton's bones.

However, this leads to a problem where we generate *too many* nodes, so that's why this PR needs to refactor the main node generation function to make this work. We need to only generate nodes when needed, only generate bone attachments when needed, only generate the correct amount of bones, and ensure nested skeleton leaf node trees stay together. All of this is required to be consistent with the behavior in Blender.

Here is a minimal reproduction asset: [skinD_09_modified.zip](https://github.com/user-attachments/files/19263440/skinD_09_modified.zip). This file is the "skinD" example in `Animation_Skin_09.gltf` which can be found here https://github.com/KhronosGroup/glTF-Asset-Generator/tree/main/Output/Positive/Animation_Skin which showcases most of the edge cases. I've modified this file to have a non-joint leaf node in the skeleton so that it showcases the problem this PR fixes. In master, this file imports like this:

Screenshot 2025-03-15 at 8 55 57 AM

With this PR, the leaf node does not get collapsed, and becomes a node, with a parent attached to the bone:

Screenshot 2025-03-17 at 1 32 16 AM

With this PR, Godot's import of the file matches the correct behavior of how Blender imports the same file.

- *Bugsquad edit:* Fixes #75917, fixes #92447

### CODE REVIEW

**Comment by fire** [comment]
There are some github action errors. They look minor.

File: modules/gltf/skin_tool.cpp [hunk 1]
```diff
@@ -403,11 +403,19 @@ Error SkinTool::_determine_skeletons(
...
 		for (int i = 0; i < skeleton_nodes.size(); ++i) {
 			const SkinNodeIndex node_i = skeleton_nodes[i];
-
-			if (nodes[node_i]->joint) {
+			Ref<GLTFNode> node = nodes[node_i];
+			if (node->joint) {
 				skeleton->joints.push_back(node_i);
-			} else {
-				non_joints.push_back(node_i);
+				// If a joint node has non-joint parents, we need to make them joints as well.
```
Comment (fire) [inline modules/gltf/skin_tool.cpp:409]: Since we're changing the algorithm for the node tree structure and that breaks the scene tree when doing inherited scenes. We've tended to do a forward upgrade import boolean.

Comment (fire) [reply modules/gltf/skin_tool.cpp:409]: It's been a long time since we've had one, but maybe @lyuma remember how we did it.

Comment (aaronfranke) [reply modules/gltf/skin_tool.cpp:409]: The set of generated nodes with this PR should be a superset of the nodes generated in master.

**Comment by fire** [comment]
We have some old tests from https://github.com/godotengine/godot-tests/tree/master/tests/gltf_skeleton_tests

File: modules/gltf/skin_tool.cpp (hunk 1 above)
Comment (fire) [reply modules/gltf/skin_tool.cpp:409]: For example inherited nodes will have a different node path. Let's say when we place a magical rod in the bone attachment. So I'm worried this will shift if the Paths are not exactly the same. I can't experiment on this due to a pending trip.

**Comment by lyuma** [comment]
If we opt to fix this, we should also fix mesh nodes. See how `transformNode` is both a node and a BoneAttachment + MeshInstance3D? Users also want those to become standalone nodes.

Our current behavior in Godot 4.x guarantees glTF compatibility, and Just Works without a sea of edge cases, but at the expense of usability for some users. As I understand it, though, your change is not fundamentally changing all aspects of the skeleton (these leaf nodes will still also be bones, but it adds a case where it is creating a bone attachment? it's a bit hard to follow what changed with the refactor)

Personally, I don't like all of the `_end` nodes creating BoneAttachments. That creates a gigantic mess on any model with hundreds of bones, and I am relieved to see those end nodes collapsed into the skeleton. I'm worried this sort of change will bring back messy skeletons, at least on FBX files (since Blender adds _end leaf bones by default).

> We need to only generate nodes when needed, only generate bone attachments when needed, only generate the correct amount of bones, and ensure nested skeleton leaf node trees stay together.

I'm not sure I understand how you can know this. I guess if it came from Blender, you can make some assumptions since Blender always explicitly exports all skeleton bones as joints (this is how you know), but in general glTF has no way of communicating if a node is "when needed". The joints array isn't guaranteed to exist (and in the leaf bone case, I don't think Blender adds those to joints during .fbx export).

> The correct behavior ...

Unless I'm mistaken, this isn't about correctness in the glTF spec: it is about matching the scene tree in Blender, right? I do like improving compatibility with Blender, but I'd be worried about introducing bugs with exports from other DCC applications (or even FBX, which also uses SkinTool)

I don't personally have the mental bandwidth to really dig into this problem, figure out which cases should be nodes and which should be bones, evaluate the edge cases, figure out animation bugs (remember that Godot stores bone and node transforms differently), and think about compatibility breakage, and then figure out how to rewrite all of my addons to support the new behavior (since having a mix of bones and nodes will break all of my skeleton modifier code). This is just too complex a problem for me to get into it for the foreseeable future... so I'd personally like to recuse myself from involvement in this PR.

The compatibility options can still be used where needed in projects that depend on the old behavior, so I understand the compatibility risk is minimal in this sense.

**Comment by aaronfranke** [comment]
> If we opt to fix this, we should also fix mesh nodes. See how transformNode is both a node and a BoneAttachment + MeshInstance3D? Users also want those to become standalone nodes.

What do you mean? The `transformNode` node is not a skinned mesh, so it must be attached to the bone. This is required to be compliant with the example file from Khronos.

> Personally, I don't like all of the _end nodes creating BoneAttachments. .. at least on FBX files since Blender adds

This does not happen with this PR, at least not with FBX files exported from Blender. `_end` bones don't get attachments.

> I'm not sure I understand how you can know this.
> Unless I'm mistaken, this isn't about correctness in the glTF spec: it is about matching the scene tree in Blender, right?

Yes, it's matching the behavior in Blender. Well, glTF doesn't specify how Godot or Blender should import these, but the existence of `joints` is pretty clearly defining which glTF nodes are bones in a skeleton.

> but in general glTF has no way of communicating if a node is "when needed"

If we don't do this, people expecting to have nodes attached to skeletons (ex: the Blender Project DogWalk team) have no way to do this... almost. If we did keep this behavior, there is still a workaround that does require the refactor in my PR but undoing the SkinTool changes. Currently Godot doesn't consider GLTFDocumentExtension classes when deciding if a bone should have a node generated, but this is a mistake. However, I don't think we should require users to do this. People should be able to attach a plain node in Blender without glTF extensions and use that in Godot directly.

> The joints array isn't guaranteed to exist

Yes, it is: https://github.com/KhronosGroup/glTF/blob/main/specification/2.0/schema/skin.schema.json#L33

> I don't personally have the mental bandwidth to really dig into this problem,

Honestly I completely understand that. It has taken me days to do this. In fact, I thought I had it, so I opened this PR, but then I found more edge cases, so I had to put this PR as draft and rework things like 3 more times before I got it right. The old test files in the repo @fire sent were immensely useful because they contained so many edge cases. I believe this PR now works correctly for all of those.

> The compatibility options can still be used where needed in projects that depend on the old behavior, so I understand the compatibility risk is minimal in this sense.

Yes, and I've updated this PR to add a new number to the glTF compatibility option.

**Comment by fire** [comment]
What's the state of this?

**Comment by aaronfranke** [comment]
@fire Probably all good now as far as I can tell, but given the impact and previous problems, we should do more testing first. Regardless I am confident this is better than the current master but ideally we find any other edge cases and fix them, if any more exist.

**Comment by fire** [comment]
Why is this still a draft pull request?

**Comment by lyuma** [comment]
I talked about this some more at the meeting. This is designed for Blender, and it makes sense for exports from DCCs that correctly make bones into joints. This change seems good in terms of allowing better round tripping between Godot and Blender.

My worry is that for glTFs exported from DCC tooling which do not have a bespoke skeleton or armature or do not use the unused skin trick to force all bones to be joints, this PR could unexpectedly cause some bones to become ordinary Node3D's.

That said, we have some control in the importer: Users who are importing a character and are dissatisfied with how this change makes some non-joint nodes into Node3D can check the "Import as Skeleton Bones" checkbox in the importer, or could lower the compatibility setting to 4.4.

I would perhaps like to see a round trip test done between Godot and itself; and Godot and Blender, to see if this change actually improves round tripping of skeleton bones and non-joint nodes.

**Comment by aaronfranke** [comment]
@lyuma Yes, this PR has the correct behavior for round-tripping out of Godot and back into itself, and into Blender.

Correctly imported Godot scene from the test file in the PR description:

Screenshot 2025-05-25 at 11 51 18 AM

Correctly exported from Godot: [exported.zip](https://github.com/user-attachments/files/20432318/exported.zip)

The mesh on `transformNode` is given its own node on import, which is the correct behavior (remember that this is an edge case), and then is given a unique name on export (not sure why `3` but again it's an edge case). The Skeleton3D's `joint3` bone is a leaf bone, which is the correct behavior. It's interesting to see that the node order changed to be alphabetical, I'm not quite sure why, but well, either order is correct. The AnimationPlayer node should probably be excluded from export, but that's a separate issue from this PR (I tested that it happens on master).

Correctly imported Godot scene from the exported glTF file:

Screenshot 2025-05-25 at 11 52 00 AM

Correctly imported into Blender:

Screenshot 2025-05-25 at 12 00 05 PM

**Comment by fire** [comment]
I think it's good to go, but will poke @lyuma for an approval.

**Comment by TokageItLab** [comment]
I am concerned about the compatibility with the Import as Skeleton Bones option implemented by https://github.com/godotengine/godot/pull/88819. Since I remember it being like making all the Armature's child nodes (even if without weight) into bones.

image

**Comment by aaronfranke** [comment]
@TokageItLab Thanks, that was indeed not working. I've updated this PR to fix that option. The new parameter added to `SkinTool::_determine_skeletons` for whether to turn non-joint descendant nodes into bones or not needs to be true in the case of importing the whole scene as skeleton bones, and true in compat mode, but false if neither of those are the case.

```cpp
if (p_state->get_import_as_skeleton_bones()) {
	err = SkinTool::_determine_skeletons(p_state->skins, p_state->nodes, p_state->skeletons, p_state->root_nodes, true);
} else {
	err = SkinTool::_determine_skeletons(p_state->skins, p_state->nodes, p_state->skeletons, Vector<GLTFNodeIndex>(), _naming_version < 2);
}
```

In a future compat breakage, we could consolidate the boolean into just checking if the root bone array is empty or not.

File: modules/gltf/gltf_document.cpp [hunk 2]
```diff
@@ -6256,11 +6265,198 @@ void GLTFDocument::_convert_mesh_instance_to_gltf(MeshInstance3D *p_scene_parent
...
+		}
+	} else {
+		// Add the node we generated and set the owner to the scene root.
+		Array args;
+		args.append(p_scene_root);
+		p_current_node->propagate_call(StringName("set_owner"), args);
+	}
+}
+
+bool GLTFDocument::_does_skinned_mesh_require_placeholder_node(Ref<GLTFState> p_state, Ref<GLTFNode> p_gltf_node) {
```
Comment (lyuma) [inline modules/gltf/gltf_document.cpp:6309]: This function causes a regression in https://github.com/godotengine/godot-tests/tree/master/tests/gltf_skeleton_tests/gltf_tests: MeshNode becomes both a bone attachment, an empty and the mesh node is renamed to MeshNode2, even though the BoneAttachment is not needed.

image
[MeshSkinnedToItselfAndOthers.glb](https://github.com/user-attachments/files/20644091/MeshSkinnedToItselfAndOthers_rename_to_glb.dmp)

To fix it, I enitrely removed `_does_skinned_mesh_require_placeholder_node` (made it always return false), and it seemed to solve the rest of the corner cases. I tested with both the above gltf_skeleton_tests as well as the MRP from [#67773](https://github.com/godotengine/godot/issues/67773#issuecomment-1374525580) which is mentioned in the "Edge case" comment below, and I found it works correctly.

**Review by lyuma — APPROVED** [review]
Seems to mostly work well for the test cases, so I think it looks good.

There's a special case function `_does_skinned_mesh_require_placeholder_node` which I think is unnecessary and incorrect. It works well after I remove it.

File: modules/gltf/gltf_document.cpp (hunk 2 above)
Comment (aaronfranke) [reply modules/gltf/gltf_document.cpp:6309]: Part of this function is an extracted version of the existing logic here https://github.com/godotengine/godot/blob/42c7f144225fa3992ceda284ad8f8edcbd1e7455/modules/gltf/gltf_document.cpp#L6322

You added this in PR #72158 actually: https://github.com/godotengine/godot/pull/72158/files#diff-fdf33f6fcd2b129d540ebd0e70fd04982b68e279cebfd9e96036d2fd7561c31fR5643

The function checks for skinned meshes (existing check), checks for non-joint children (modified from the check for no children), and checks if child joints don't have their skeletons in the tree yet (new check, an edge case that becomes required to check for because of the modified second check). Replacing the function with `return false;` ignores part of the point of this PR, so that's the wrong move.

I think this explains it: The function is technically correct that we do need a placeholder in this case, but where we go wrong is by generating a new placeholder node. The subtlety here is that in this case, the Skeleton3D itself should be the placeholder. In some cases we need a new placeholder within a skeleton, in most cases we need to make a new Skeleton3D node that doesn't map to any glTF node, but in this case, these overlap, so we can feed two birds with one scone by using the Skeleton3D as the placeholder, inserting it into the tree immediately.

I just pushed this PR with this diff, which fixes the file you linked.

Screenshot 2025-06-08 at 2 14 26 PM

Additionally, this now means the generated tree matches what `GLTFNode.get_scene_node_path` expects with PR https://github.com/godotengine/godot/pull/104787 included (errors in the current master).

EDIT: It is also possible that user code generates a Skeleton3D node in a GLTFDocumentExtension, and this code also works for that case, so long as the user wants it named Skeleton3D or sets its name manually.

**Review by lyuma — APPROVED** [review]
It's hard for me to conceptualize enough of this code to be 100% certain, but you offer a reasonable explanation of the fix to the placeholder -> skinned mesh issue, and it seems to cover the test cases I have, so I'm satisfied.

I'll continue testing during beta, but I think it's good to merge. Nice work solving such a difficult round trip problem, Aaron.

**Comment by Repiteo** [comment]
Thanks!

**Comment by JonqsGames** [comment]
Sorry but the simple test project on my PR #98516 is not working as expected. I think it even worse since now parenting looks mixed up. Can you confirm that this PR was meant to fix the leaf bone created by godot when mesh is parented to a bone ?

**Comment by aaronfranke** [comment]
@JonqsGames Make sure you change the naming version to "Godot 4.5 or later" in the Import dock.
